proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -756,136 +717,178 @@ lemma setEndpoint_pde_mappings'[wp]:
+  apply (drule(1) ep_ko_at_valid_objs_valid_ep')
+  apply (rename_tac ko a list)
+  apply (prop_tac "ko \<noteq> IdleEP \<longrightarrow> valid_ep' (epQueue_update (\<lambda>_. a # list) ko) s")
+   apply (clarsimp simp: nonempty_epQueue_remove1_valid_ep')
+  apply fastforce
+  done
+
+lemma cancelIPC_sch_act_wf[wp]:
+  "\<lbrace>\<lambda>s. sch_act_wf (ksSchedulerAction s) s \<and> sch_act_not tptr s\<rbrace>
+   cancelIPC tptr
+   \<lbrace>\<lambda>_ s. sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  unfolding cancelIPC_def
+  apply (wpsimp wp: gts_wp' hoare_vcg_imp_lift' threadSet_sch_act hoare_vcg_all_lift
+                    replyRemoveTCB_sch_act_wf)
+  done
+
+(* FIXME RT: move to ...? *)
+crunches getBlockingObject
+  for inv: P
+
+(* FIXME RT: move to...? *)
+lemma endpoint_live:
+  "\<lbrakk>ko_at' ep ptr s; ep \<noteq> IdleEP\<rbrakk> \<Longrightarrow> ko_wp_at' live' ptr s"
+  apply (clarsimp simp: ko_wp_at'_def obj_at'_def projectKOs)
+  done
+
+lemma blockedCancelIPC_if_live'[wp]:
+  "blockedCancelIPC st tptr epptr \<lbrace>if_live_then_nonz_cap'\<rbrace>"
+  unfolding blockedCancelIPC_def getBlockingObject_def
+  apply (wpsimp wp: getEndpoint_wp haskell_assert_wp)
+  apply (clarsimp simp: if_live_then_nonz_cap'_def endpoint.disc_eq_case endpoint_live)
+  done
+
+lemma blockedCancelIPC_valid_idle':
+  "\<lbrace>valid_idle' and valid_pspace' and (\<lambda>s. tptr \<noteq> ksIdleThread s)\<rbrace>
+   blockedCancelIPC st tptr epptr
+   \<lbrace>\<lambda>_. valid_idle'\<rbrace>"
+  unfolding blockedCancelIPC_def getBlockingObject_def
+  apply (wpsimp wp: getEndpoint_wp)
+  apply (case_tac "remove1 tptr (epQueue ko)"; clarsimp; normalise_obj_at')
+   apply (clarsimp simp: valid_ep'_def)
+  apply (drule(1) ep_ko_at_valid_objs_valid_ep'[OF _ valid_pspace_valid_objs'])
+  apply (rename_tac ko a list)
+  apply (prop_tac "ko \<noteq> IdleEP \<longrightarrow> valid_ep' (epQueue_update (\<lambda>_. a # list) ko) s")
+   apply (clarsimp simp: nonempty_epQueue_remove1_valid_ep')
+  apply fastforce
+  done
+
+crunches blockedCancelIPC
+  for ct_not_inQ[wp]: ct_not_inQ
+  and cur_tcb'[wp]: "cur_tcb'"
+  and ctes_of[wp]: "\<lambda>s. P (ctes_of s)"
+  and valid_queues'[wp]: valid_queues'
+  and valid_release_queue[wp]: valid_release_queue
+  and valid_release_queue'[wp]: valid_release_queue'
+  and untyped_ranges_zero'[wp]: "untyped_ranges_zero'"
+  (wp: crunch_wps)
 
-lemma replyUnlink_sch_act_simple[wp]:
-  "replyUnlink r t' \<lbrace>sch_act_simple\<rbrace>"
-  apply (simp add: replyUnlink_def setReplyTCB_def getReplyTCB_def)
-  by (wpsimp wp: hoare_drop_imps)
+(* FIXME RT: move to... ?*)
+lemma add_asserts:
+  assumes not: "\<lbrace>Not o P\<rbrace> f \<lbrace>\<lambda>_. \<bottom>\<rbrace>"
+  shows "\<lbrace>\<lambda>s. P s \<and> Q s\<rbrace> f \<lbrace>R\<rbrace> \<Longrightarrow> \<lbrace>Q\<rbrace> f \<lbrace>R\<rbrace>"
+  apply (insert use_valid[OF _ not])
+  apply (fastforce simp: valid_def)
+  done
+
+lemma blockedCancelIPC_asserts:
```

Comment:
We probably don’t need to have this as a separate lemma. Is there any particular reason you cannot include this in the main proof below?